src/main/java/com/google/sps/lib/TSPDynamicProgrammingAlgorithm.java

```diff
@@ -14,15 +14,89 @@
 
 package com.google.sps.lib;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 
 /**
  * Implements TSP algorithm in exponential time using dynamic programming.
  */
 public class TSPDynamicProgrammingAlgorithm implements ITSPAlgorithm {
 
-    @Override
-    public List<String> findShortestPath(double [][]distanceMatrix) {
-        return null;
+  @Override
+  public List<String> findShortestPath(double[][] distanceMatrix) {
+    Integer[][] path = new Integer[(1 << distanceMatrix.length)][distanceMatrix[0].length];
+    double[][] dp = new double[(1 << distanceMatrix.length)][distanceMatrix[0].length];
+    for (int i = 0; i < dp.length; i++) {
+      Arrays.fill(dp[i], Integer.MAX_VALUE);
     }
+    
+    if (!checkNegativeDistance(distanceMatrix)) {
+      tsp(distanceMatrix, 1, 0, path, dp);
+      return getTour(path);
+    }
+    return null;
+  }
+
+  private double tsp(double[][] distanceMatrix, int stage, int pos, Integer[][] path, double[][] dp) {
+    if (stage == (1 << distanceMatrix.length) - 1) {
+      return distanceMatrix[pos][0];
+    }
+    if (dp[stage][pos] != Integer.MAX_VALUE) {
+      return dp[stage][pos];
+    }
+
+    double ans = Double.MAX_VALUE;
```

Comment:
Okay. One commment for the entire PR. Please keep the identifier naming verbose. What is ans supposed to mean?